devito/interfaces.py

```diff
@@ -440,44 +445,126 @@ class DenseData(TensorData):
     def __init__(self, *args, **kwargs):
         if not self._cached():
             self.name = kwargs.get('name')
-            self.shape = kwargs.get('shape', None)
-            if self.shape is None:
-                dimensions = kwargs.get('dimensions')
-                self.shape = tuple([d.size for d in dimensions])
+            self.grid = kwargs.get('grid', None)
+
+            if self.grid is None:
+                self.shape_domain = kwargs.get('shape', None)
+                self.dtype = kwargs.get('dtype', np.float32)
+                if self.shape_domain is None:
+                    error("Creating a Function requires either 'shape'"
+                          "or a 'grid' argument")
+                    raise ValueError("Unknown symbol dimensions or shape")
+            else:
+                self.shape_domain = self.grid.shape_domain
+                self.dtype = kwargs.get('dtype', self.grid.dtype)
             self.indices = self._indices(**kwargs)
-            self.dtype = kwargs.get('dtype', np.float32)
+
             self.space_order = kwargs.get('space_order', 1)
             self.initializer = kwargs.get('initializer', None)
             if self.initializer is not None:
                 assert(callable(self.initializer))
             self._first_touch = kwargs.get('first_touch', configuration['first_touch'])
             self._data_object = None
 
+            # Dynamically create notational shortcuts for space derivatives
+            for dim in self.space_dimensions:
+                # First derivative, centred
+                dx = partial(first_derivative, order=self.space_order,
+                             dim=dim, side=centered)
+                setattr(self.__class__, 'd%s' % dim.name,
+                        property(dx, 'Symbol for the first derivative '
+                                 'wrt the %s dimension' % dim.name))
+
+                # First derivative, left
+                dxl = partial(first_derivative, order=self.space_order,
+                              dim=dim, side=left)
+                setattr(self.__class__, 'd%sl' % dim.name,
+                        property(dxl, 'Symbol for the first derivative '
+                                 'wrt %s with a left stencil' % dim.name))
```

Comment:
for homogeneity :
```
Symbol for the first derivative wrt the %s dimension with a left stencil
```
Perhaps more correct
```
Return the symbolic first-order left-sided approximation wrt the %s dimension
```
Note that in the case of a centered approximation the approximation is second order

Reply:
and, probably, also the other automatically generated docstrings below could be tweaked a bit